pkg/lib/synthesize/resource_getters.go

```diff
@@ -0,0 +1,301 @@
+	ServiceMonitorKind     = "ServiceMonitor"
+	DeploymentKind         = "Deployment"
+	APIServiceKind         = "APIService"
+)
+
+// registerGVK registers the GVK of the kube resource we use in the CSV, explicitly including all versions.
+func registerGVK(scheme *runtime.Scheme) {
+	scheme.AddKnownTypes(operatorsv1alpha1.SchemeGroupVersion, &operatorsv1alpha1.ClusterServiceVersion{})
+
+	scheme.AddKnownTypes(extv1.SchemeGroupVersion, &extv1.CustomResourceDefinition{})
+	scheme.AddKnownTypes(extv1beta1.SchemeGroupVersion, &extv1beta1.CustomResourceDefinition{})
+
+	scheme.AddKnownTypes(cv1.SchemeGroupVersion,
+		&cv1.ServiceAccount{},
+		&cv1.ConfigMap{},
+		&cv1.Secret{})
+
+	scheme.AddKnownTypes(rv1.SchemeGroupVersion,
+		&rv1.Role{},
+		&rv1.RoleBinding{},
+		&rv1.ClusterRole{},
+		&rv1.ClusterRoleBinding{})
+
+	scheme.AddKnownTypes(appsv1.SchemeGroupVersion, &appsv1.Deployment{})
+
+	scheme.AddKnownTypes(apiregistrationv1.SchemeGroupVersion, &apiregistrationv1.APIService{})
+	scheme.AddKnownTypes(apiregistrationv1beta1.SchemeGroupVersion, &apiregistrationv1beta1.APIService{})
+}
+
+type GroupVersionKindName struct {
+	Group   string
+	Version string
+	Kind    string
+	Name    string
+}
+
+func NewGVKN(gvk schema.GroupVersionKind, name string) GroupVersionKindName {
+	return GroupVersionKindName{
+		Group:   gvk.Group,
+		Version: gvk.Version,
+		Kind:    gvk.Kind,
+		Name:    name,
+	}
+}
+
+func (gvkn *GroupVersionKindName) gvk() GroupVersionKind {
+	return GroupVersionKind{
+		Group:   gvkn.Group,
+		Version: gvkn.Version,
+		Kind:    gvkn.Kind,
+	}
+}
+
+type GroupVersionKind schema.GroupVersionKind
+
+type customResourceDefinition struct {
+	v1      extv1.CustomResourceDefinition
+	v1beta1 extv1beta1.CustomResourceDefinition
+	gvkn    GroupVersionKindName
+}
+
+func (crd *customResourceDefinition) getResourceGVKNs() (gvkns []GroupVersionKindName) {
+	switch crd.gvkn.Version {
+	case crd.v1.GroupVersionKind().Version:
+		for _, version := range crd.v1.Spec.Versions {
+			gvkns = append(gvkns, GroupVersionKindName{
+				Group:   crd.v1.Spec.Group,
+				Kind:    crd.v1.Spec.Names.Kind,
+				Version: version.Name,
+				Name:    crd.v1.GetName(),
+			})
+		}
+	case crd.v1beta1.GroupVersionKind().Version:
+		for _, version := range crd.v1beta1.Spec.Versions {
+			gvkns = append(gvkns, GroupVersionKindName{
+				Group:   crd.v1beta1.Spec.Group,
+				Kind:    crd.v1beta1.Spec.Names.Kind,
+				Version: version.Name,
+				Name:    crd.v1beta1.GetName(),
+			})
+		}
+	}
+	return
+}
+
+func (s *synthesize) getCRDs() ([]customResourceDefinition, error) {
+	if crdIFs, ok := s.structuredObjectMap[CRDKind]; ok {
+		var crds []customResourceDefinition
+		for _, bufObj := range crdIFs {
+			switch bufObj.GroupVersionKind().GroupVersion() {
+			case extv1.SchemeGroupVersion:
+				crd := extv1.CustomResourceDefinition{}
+				if err := runtime.DefaultUnstructuredConverter.FromUnstructured(bufObj.UnstructuredContent(), &crd); err != nil {
```

Comment:
correct me if I'm wrong @ecordell but I believe we ran into issues unmarshaling CRDs via unstructured (if I'm understanding this code correctly). There is a known int -> float conversion error among others. I think we should write directly to the type via something like https://github.com/operator-framework/operator-lifecycle-manager/blob/9ffa1fdb8dcf3153f915563641dc4ceaf3fbe3f2/pkg/lib/crd/unmarshal.go#L29. 

But I'm not sure if I have the full context here, maybe this is fine

Reply:
I was not aware of the conversion problem. we can change it.

Reply:
Can we have a confirmation on this issue? We do use the `unstructured` in a few places. I'm not aware of this conversion error.

Reply:
Does it make sense that we are not applying the CRD but to get the resource GVK that this should be fine?

Reply:
Yes, in this case it should be fine. I've seen this issue specifically around schema validation. As long as we are not stepping through unstructured to get to a CRD serialized type that we apply to a cluster, we should be fine.